### DISCUSSION TITLE
Use updated edge speeds for route time calculation

### DISCUSSION THREAD

**Discussion by mohit-at-delhivery** [Project-OSRM/osrm-backend#7021] [Q&A]
Hi everyone,

I have a osm .pub file from geofabrik. I want to know which tags (for nodes and ways) should i update in this osm file using osmium so that once we have extracted and partitioned osrm from this osm file lets say using car.lua, then upon calling the route api i would like to see the updated duration when calculating route time between two points.  

There was an issue which also asked which speed is taken into consideration when calculating route duration from route api (issue: 6872): [Link](https://github.com/Project-OSRM/osrm-backend/issues/6872)

**Comment by jcoupey** [comment]
Rather than modifying the pbf file prior to extracting with OSRM, you may consider adjusting speeds afterward using [traffic updates](https://github.com/Project-OSRM/osrm-backend/wiki/Traffic).

**Reply by mohit-at-delhivery** [reply]
@jcoupey Thank you so much for referring the traffic updates approach. Will look onto that as well. Do you know what will be the difference in the two approaches?

Base: We have 'n' ways for which we want to update the duration. For each way out of these n ways we have a custom speed.

1. We update the maxspeed in the osm data using our custom speed and then extract it and run osrm on it.
2. We use the csv file containing the custom speeds and use --generate-edge-lookup while extracting osrm as suggested.

Now when we calculate route duration using osrm route api between same two points, then are we going to get the same output?

**Comment by mohit-at-delhivery** [comment]
@jcoupey I am trying to include the traffic updates in my osrm, but i am not seeing any change in the response of the route api. I have a small area osm file. I have been using the following commands to set up osrm on my machine since a long time and it works fine. Now the only updates to include segment speeds are in steps 1 and 3, where in step 1 we are adding --generate-edge-lookup and in step 3 where we are adding the csv file for updates.

```
### OSRM Extract
docker run -t -v $(pwd):/data ghcr.io/project-osrm/osrm-backend:v5.27.1 osrm-extract -p /opt/car.lua /data/my_map.pbf --generate-edge-lookup

### Partion this graph recursively into cells
docker run -t -v $(pwd):/data ghcr.io/project-osrm/osrm-backend:v5.27.1 osrm-partition /data/my_map.osrm

### Customize the cells by calculating routing weights for cells
docker run -t -v $(pwd):/data ghcr.io/project-osrm/osrm-backend:v5.27.1 osrm-customize /data/my_map.osrm --segment-speed-file /data/custom_segments_speed.csv

### Run a docker for inferencing
docker run -t -i -p 5003:5000 -v $(pwd):/data -d --name segment_speed_map_container ghcr.io/project-osrm/osrm-backend:v5.27.1 osrm-routed --thread 14 --algorithm mld --max-table-size 100000000 --max-trip-size 100000000 /data/my_map.osrm

```

In the custom_segments_speed.csv, i have set 1000 km/hr speed for know segments still i am not able to see any impact on the route api.

Am i missing something? Also what can i do to take care of this point mentioned in traffic updates page:

"Remember to re-load the data by either restarting osrm-routed, or using osrm-routed -s along with osrm-datastore."

Any help will be appreciated. Thank you.

**Reply by mohit-at-delhivery** [reply]
do we have to change something in car.lua as well? please help

**Reply by mohit-at-delhivery** [reply]
@jcoupey It is working (in the sense that the segment speed values are different from the original osrm) when i used more sample points in csv and without --generate-edge-lookup in step 1. 

However when i am hitting the route API and then checking the segment (node A to node B) speed and duration, they are no where near to what i had inserted in the csv for the same node pair. I have simply created the csv where each row has: from node, to node, new_speed in km/h, <empty>. So that existing edge weight might be kept. Also a bit confused between the differences in edge weight and duration and their relation with the new edge speed.

Another concern i am having even without updating osrm is that i am getting different route time using route api vs table api, though the distances are exactly same.


geo_coords = ((28.4764588, 77.012971), (28.4764649, 77.015579))


# OSRM Route API Request
http://localhost:5001/route/v1/driving/77.012971,28.4764588;77.015579,28.4764649?overview=full&geometries=geojson&annotations=true

total_time:  47.6 seconds
total_distance:  330.345101321 m

# OSRM Table API Request
http://127.0.0.1:5001/table/v1/driving/77.012971,28.4764588;77.015579,28.4764649?annotations=distance,duration

distance_matrix: array([[  0. , 330.3],
        [330.3,   0. ]])

 time_matrix: array([[ 0. , 62.3],
        [62.2,  0. ]])

Please let me know how can i get the same custom segment speed in the output and also which api should i use route vs table if my aim is to only predict the new travel time between any two locations.

**Reply by mohit-at-delhivery** [reply]
@danpat could you please help here? Like i have tried to put 1000 km/hr in some segments but when i see the speed for the segment from the route api response its randomly in the range from 150 to 300 km/hr. Can i do something to get my exact speed for the custom segment in the response as well and the duration based on my speed only? and also regarding the above route api time. Is the route api not considering the turn penalty or something? Thank you.
